README.md
# WebFX

A configurable shim for running a web application in a local JavaFX
Webkit web view.

## Usage

Start your webserver, pass a configuration map (`Map<String, String>`)
to the static method `webfx.WebFX.start`.

Important options:

- `windowTitle` - The text to set the window title to - Defaults to `WebFX`
- `domain` - When combined with the port, determines the location of the
  page to show. Defaults to `localhost`
- `port` - When combined with the domain, determinse the location of
  the page to show. Defaults to `3000`
- `width` - The width of the JavaFX window. Defaults to `1024`
- `height` - The height of the JavaFX window. Defaults to `768`

## Requirements

Needs to be compiled _and_ run with Java 8 or later.

## License

Copyright © 2017 Geoff Shannon

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this
file. This change log follows the conventions
of [keepachangelog.com](http://keepachangelog.com/).


src/webfx/App.java
package webfx;

import java.util.ArrayList;
import java.util.Map;
import java.util.Map.Entry;
import java.util.List;

import javafx.application.Application;
import javafx.scene.Scene;
import javafx.scene.web.WebView;
import javafx.scene.web.WebEngine;
import javafx.stage.Stage;

public class App extends Application {
    private String windowTitle;

    private String domain;
    private String port;

    private int width;
    private int height;

    private int getIntArgOrDefault(Map<String, String> args, String argName, int defaultValue) {
        try {
            return Integer.parseInt(args.getOrDefault(argName, ""));
        } catch (NumberFormatException e) {
            return defaultValue;
        }
    }

    @Override
    public void init() throws Exception {
        Map<String, String> args = getParameters().getNamed();

        this.windowTitle = args.getOrDefault("windowTitle", "Java WebFX Application");

        this.domain = args.getOrDefault("domain", "localhost");
        this.port   = args.getOrDefault("port", "3000");

        this.width  = getIntArgOrDefault(args, "width", 1024);
        this.height = getIntArgOrDefault(args, "height", 768);
    }

    @Override
    public void start(Stage primaryStage) {
        primaryStage.setTitle(this.windowTitle);

        WebView browser = new WebView();
        WebEngine engine = browser.getEngine();

        engine.load(String.format("http://%s:%s", this.domain, this.port));

        Scene s = new Scene(browser, this.width, this.height);

        primaryStage.setScene(s);
        primaryStage.show();
    }

    private static String[] constructArgVector(Map<String, Object> options) {
        List<String> args = new ArrayList<String>();

        for (Entry<String, Object> option: options.entrySet()) {
            String argName  = option.getKey();
            String argValue = option.getValue().toString();
            args.add(String.format("--%s=%s", argName, argValue));
        }

        return args.toArray(new String[0]);
    }

    public static void start(Map<String, Object> options) {
        String[] args = constructArgVector(options);
        launch(args);
    }

    public static void main(String[] args) {
        launch(args);
    }

}


src/webfx/Server.java
package webfx;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

import java.net.InetSocketAddress;
import java.net.URI;

import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;

import java.util.Arrays;

import java.util.stream.Collectors;
import java.util.stream.Stream;

import com.sun.net.httpserver.Headers;
import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpHandler;
import com.sun.net.httpserver.HttpServer;

public class Server {

    public static void main(String[] args) throws Exception {
        System.out.println("Starting web server on port 3000...");
        HttpServer server = HttpServer.create(new InetSocketAddress(3000), 0);
        server.createContext("/", new Handler(Paths.get(".")));
        server.start();
    }

    static class Handler implements HttpHandler {
        private static final String HEADER = "<!DOCTYPE html>\n<html><head></head><body>";
        private static final String FOOTER = "</body></html>";

        private final Path root;

        public Handler(Path root) {
            this.root = root;
        }

        private Path relativeToRoot(Path branch) {
            return this.root.relativize(branch);
        }

        public String getDirectoryContent(Path content) throws IOException {
            Stream<String> headings = Stream.of(String.format("<h1>Index of /%s</h1>",
                                                              relativeToRoot(content).getFileName()));

            Path parentPath = content.getParent();
            if (parentPath != null && parentPath.startsWith(this.root)) {
                headings = Stream.concat(headings,
                                         Stream.of(String.format("<a href=\"/%s\">..</a><br>",
                                                                 relativeToRoot(parentPath).toString())));
            }

            return
                Stream.concat(headings,
                              Files.list(content)
                              .map((Path entry) -> String.format("<a href=\"/%s\">%s</a><br>",
                                                                 relativeToRoot(entry).toString(),
                                                                 entry.getFileName())))
                .collect(Collectors.joining("", HEADER, FOOTER));
        }

        private void copy(InputStream is, OutputStream outStream) throws IOException {
            try (BufferedInputStream inStream  = new BufferedInputStream(is)) {
                int buffSize = 8 * 1024;
                byte[] buffer = new byte[buffSize];

                int bytesRead = inStream.read(buffer, 0, buffSize);
                while (bytesRead != -1) {
                    outStream.write(buffer, 0, bytesRead);
                    bytesRead = inStream.read(buffer, 0, buffSize);
                }

                outStream.flush();
            }
        }

        @Override
        public void handle(HttpExchange t) throws IOException {
            URI uri = t.getRequestURI();

            Path content = Paths.get(uri.toString());
            Path relativeRequestPath = Paths.get("/").relativize(content);
            Path absolutePath = this.root.resolve(relativeRequestPath);

            String contentType = "";
            InputStream response = null;
            long responseSize = 0;

            try {
                if (Files.isRegularFile(absolutePath)) {
                    contentType = "text/plain";
                    response = Files.newInputStream(absolutePath);
                    responseSize = Files.size(absolutePath);

                } else if (Files.isDirectory(absolutePath)) {
                    contentType = "text/html";
                    byte[] bytes = getDirectoryContent(absolutePath).getBytes();
                    response = new ByteArrayInputStream(bytes);
                    responseSize = bytes.length;

                } else {
                    contentType = "text/plain";
                    response = new ByteArrayInputStream(new byte[0]);
                    responseSize = 0;
                }

            } catch (Throwable e) {
                e.printStackTrace();
            }

            Headers headers = t.getResponseHeaders();
            headers.put("Content-Type", Arrays.asList(contentType));

            t.sendResponseHeaders(200, responseSize);

            OutputStream os = t.getResponseBody();
            copy(response, os);
            os.close();
        }
    }
}


src/radicalzephyr/WebFX.java
package radicalzephyr;

import java.util.ArrayList;
import java.util.Map;
import java.util.Map.Entry;
import java.util.List;

import javafx.application.Application;
import javafx.scene.Scene;
import javafx.scene.web.WebView;
import javafx.scene.web.WebEngine;
import javafx.stage.Stage;

public class WebFX extends Application {
    private String windowTitle;

    private String domain;
    private String port;

    private int width;
    private int height;

    private int getIntArgOrDefault(Map<String, String> args, String argName, int defaultValue) {
        try {
            return Integer.parseInt(args.getOrDefault(argName, ""));
        } catch (NumberFormatException e) {
            return defaultValue;
        }
    }

    @Override
    public void init() throws Exception {
        Map<String, String> args = getParameters().getNamed();

        this.windowTitle = args.getOrDefault("windowTitle", "Java WebFX Application");

        this.domain = args.getOrDefault("domain", "localhost");
        this.port   = args.getOrDefault("port", "3000");

        this.width  = getIntArgOrDefault(args, "width", 1024);
        this.height = getIntArgOrDefault(args, "height", 768);
    }

    @Override
    public void start(Stage primaryStage) {
        primaryStage.setTitle(this.windowTitle);

        WebView browser = new WebView();
        WebEngine engine = browser.getEngine();

        engine.load("http://" + this.domain + ":" + this.port);

        Scene s = new Scene(browser, this.width, this.height);

        primaryStage.setScene(s);
        primaryStage.show();
    }

    private static String[] constructArgVector(Map<String, Object> options) {
        List<String> args = new ArrayList<String>();

        for (Entry<String, Object> option: options.entrySet()) {
            String argName  = option.getKey();
            String argValue = option.getValue().toString();
            args.add(String.format("--%s=%s", argName, argValue));
        }

        return args.toArray(new String[0]);
    }

    public static void start(Map<String, Object> options) {
        String[] args = constructArgVector(options);
        launch(args);
    }

    public static void main(String[] args) {
        launch(args);
    }

}
